File: jms-benchmark/src/main/scala/org/fusesource/jmsbenchmark/JMSClientScenario.scala
/**
 * Copyright (C) 2009-2011 the original author or authors.
 * See the notice.md file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.fusesource.jmsbenchmark

import java.lang.Thread
import javax.jms._
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.CyclicBarrier

/**
 * <p>
 * Simulates load on a JMS sever using the JMS messaging API.
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
abstract class JMSClientScenario extends Scenario {

  var allow_worker_interrupt = false

  def createProducer(i:Int) = {
    new ProducerClient(i)
  }
  def createConsumer(i:Int) = {
    new ConsumerClient(i)
  }

  protected def destination(i:Int):Destination

  def indexed_destination_name(i:Int) = destination_type match {
    case "queue" => queue_prefix+destination_name+"-"+(i%destination_count)
    case "topic" => topic_prefix+destination_name+"-"+(i%destination_count)
    case _ => sys.error("Unsuported destination type: "+destination_type)
  }


  protected def factory:ConnectionFactory

  def jms_ack_mode = {
    ack_mode match {
      case "auto" => Session.AUTO_ACKNOWLEDGE
      case "client" => Session.CLIENT_ACKNOWLEDGE
      case "dups_ok" => Session.DUPS_OK_ACKNOWLEDGE
      case "transacted" => Session.SESSION_TRANSACTED
      case _ => throw new Exception("Invalid ack mode: "+ack_mode)
    }
  }

  def load_start_rendezvous(client:JMSClient, session:Session) = {
    if ( connected_latch !=null ) {
      connected_latch.countDown()
    }
    if ( apply_load_latch !=null ) {
      apply_load_latch.await()
    }
  }

  trait JMSClient extends Client {

    @volatile
    var connection:Connection = _
    var message_counter=0L
    val done = new AtomicBoolean()

    var worker:Thread = _

    var close_thread:Thread = null
    def dispose {
      if( close_thread==null ) {
        close_thread = new Thread(name+" closer") {
          override def run() {
            try {
              connection.close()
            } catch {
              case _ =>
            }
          }
        }
        close_thread.start()
      }
    }

    def execute:Unit

    def start = {
      worker = new Thread(name+" worker") {
        override def run() {
          var reconnect_delay = 0
          while( !done.get ) {
            try {

              if( reconnect_delay!=0 ) {
                Thread.sleep(reconnect_delay)
                reconnect_delay=0
              }
              connection = factory.createConnection(user_name, password)
              if( durable ) {
                connection.setClientID(name)
              }
              connection.setExceptionListener(new ExceptionListener {
                def onException(exception: JMSException) {
                }
              })
              connection.start()

              execute

            } catch {
              case e:Throwable =>
                if( !done.get ) {
                  if( display_errors ) {
                    e.printStackTrace
                  }
                  error_counter.incrementAndGet
                  reconnect_delay = 1000
                }
            } finally {
              dispose
            }
          }
        }
      }
      worker.start
    }

    def shutdown = {
      done.set(true)
      dispose
    }

    def wait_for_shutdown = {
      if ( worker!=null ) {
        val start = System.currentTimeMillis();
        worker.join(6000)
        while(worker.isAlive ) {
          if( allow_worker_interrupt ) {
            println(name+" worker did not shutdown quickly.. interrupting thread.")
            worker.interrupt()
            worker.join(1000)
          } else {
            println(name+" worker did not shutdown quickly...")
            worker.join(1000)
          }
        }
        if( close_thread!=null ) {
          val wait = 6000 - (System.currentTimeMillis() - start)
          if( wait > 0 ) {
            close_thread.join(wait)
          }
          while(close_thread.isAlive ) {
            if( allow_worker_interrupt ) {
              println(name+" closing thread did not shutdown quickly.. interrupting thread.")
              close_thread.interrupt()
              close_thread.join(1000)
            } else {
              println(name+" closing thread did not shutdown quickly...")
              close_thread.join(1000)
            }
          }
          close_thread.join()
        }
        worker = null
      }
    }
    def name:String
  }

  class ConsumerClient(override val id: Int) extends JMSClient {
    def name = "consumer " + id

    def execute {
      var session = if(tx_size==0) {
        connection.createSession(false, jms_ack_mode)
      } else {
        connection.createSession(true, Session.SESSION_TRANSACTED)
      }

      var consumer = if( durable ) {
        session.createDurableSubscriber(destination(id).asInstanceOf[Topic], name, selector, no_local)
      } else {
        session.createConsumer(destination(id), selector, no_local)
      }

      load_start_rendezvous(this, session)

      var tx_counter = 0
      while( !done.get() ) {
        val msg = consumer.receive(500)
        if( msg!=null ) {
          val latency = System.nanoTime() - msg.getLongProperty("ts")
          update_max_latency(latency)
          consumer_counter.incrementAndGet()

          val sleep  = _consumer_sleep(this)
          if (sleep != 0) {
            Thread.sleep(sleep)
          }
          if(session.getAcknowledgeMode == Session.CLIENT_ACKNOWLEDGE) {
            msg.acknowledge();
          }

          if ( tx_size != 0 ) {
            tx_counter += 1
            if ( tx_counter == tx_size) {
              session.commit()
              tx_counter = 0
            }
          }

        } else {
          // commit early once we don't get anymore messages.
          if ( tx_size != 0 && tx_counter > 0 ) {
            session.commit()
            tx_counter = 0
          }
        }

      }
    }

  }

  class ProducerClient(override val id: Int) extends JMSClient {

    def name = "producer " + id

    def execute {

      var session = if(tx_size==0) {
        connection.createSession(false, jms_ack_mode)
      } else {
        connection.createSession(true, Session.SESSION_TRANSACTED)
      }

      val producer:MessageProducer = session.createProducer(destination(id))
      producer.setDeliveryMode(if( persistent ) {
        DeliveryMode.PERSISTENT
      } else {
        DeliveryMode.NON_PERSISTENT
      })

      val msg = session.createTextMessage(body(name))
      headers_for(id).foreach { case (key, value) =>
        msg.setStringProperty(key, value)
      }

      load_start_rendezvous(this, session)

      var tx_counter = 0
      while( !done.get() ) {
        msg.setLongProperty("ts", System.nanoTime());
        producer.send(msg)
        producer_counter.incrementAndGet()

        val sleep = _producer_sleep(this)
        if (sleep != 0) {
          Thread.sleep(sleep)
        }

        if ( tx_size != 0 ) {
          tx_counter += 1
          if ( tx_counter == tx_size) {
            session.commit()
            tx_counter = 0
          }
        }
      }

    }
  }

  def body(name:String) = {
    val buffer = new StringBuffer(message_size)
    buffer.append("Message from " + name+"\n")
    for( i <- buffer.length to message_size ) {
      buffer.append(('a'+(i%26)).toChar)
    }
    var rc = buffer.toString
    if( rc.length > message_size ) {
      rc.substring(0, message_size)
    } else {
      rc
    }
  }



}

File: jms-benchmark/src/main/java/org/fusesource/jmsbenchmark/JMSClientScenario.java
/**
 * Copyright (C) 2009-2015 the original author or authors.
 * See the notice.md file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.fusesource.jmsbenchmark;

import javax.jms.*;
import java.util.concurrent.atomic.AtomicBoolean;

abstract public class JMSClientScenario extends Scenario {


    Client createProducer(int i) {
        return new ProducerClient(i);
    }

    Client createConsumer(int i, boolean drain) {
        return new ConsumerClient(i, drain);
    }

    abstract protected Destination destination(int i) throws Exception;

    String indexedDestinationName(int i) {
        if ("queue".equals(destinationType)) {
            return queuePrefix + getDestinationName() + "-" + (i % destinationCount);
        }
        if ("topic".equals(destinationType)) {
            return topicPrefix + getDestinationName() + "-" + (i % destinationCount);
        }
        throw new RuntimeException("Unsupported destination type: " + destinationType);
    }

    abstract protected ConnectionFactory factory() throws Exception;

    int jmsAckMode() {
        if ("auto".equals(ackMode)) {
            return Session.AUTO_ACKNOWLEDGE;
        }
        if ("client".equals(ackMode)) {
            return Session.CLIENT_ACKNOWLEDGE;
        }
        if ("dups_ok".equals(ackMode)) {
            return Session.DUPS_OK_ACKNOWLEDGE;
        }
        if ("transacted".equals(ackMode)) {
            return Session.SESSION_TRANSACTED;
        }
        throw new RuntimeException("Invalid ack mode: " + ackMode);
    }

    void loadStartRendezvous(JMSClient client, Session session) throws InterruptedException {
        if (connectedLatch != null) {
            connectedLatch.countDown();
        }
        if (applyLoadLatch != null) {
            applyLoadLatch.await();
        }
    }

    abstract class JMSClient implements Client {

        volatile
        Connection connection;
        long messageCounter = 0L;
        AtomicBoolean done = new AtomicBoolean();

        Thread worker;

        Thread closeThread = null;

        synchronized void dispose() {
            if (this.connection != null) {
                final Connection connection = this.connection;
                closeThread = new Thread(name() + " closer") {
                    @Override
                    public void run() {
                        try {
                            connection.close();
                        } catch (Throwable e) {
                        }
                    }
                };
                closeThread.start();
                this.connection = null;
            }
        }

        abstract void execute() throws JMSException, InterruptedException, Exception;

        public void start() {
            worker = new Thread(name + " worker") {
                @Override
                public void run() {
                    int reconnectDelay = 0;
                    while (!done.get()) {
                        try {

                            if (reconnectDelay != 0) {
                                Thread.sleep(reconnectDelay);
                                reconnectDelay = 0;
                            }

                            // Don't reconnect until we finish disconnecting..
                            if (closeThread != null) {
                                closeThread.join();
                                closeThread = null;
                            }

                            connection = factory().createConnection(userName, password);
                            if (durable) {
                                connection.setClientID(name);
                            }
                            connection.setExceptionListener((e) -> {
                            });
                            connection.start();

                            execute();

                        } catch (Throwable e) {
                            if (!done.get()) {
                                if (displayErrors) {
                                    e.printStackTrace();
                                }
                                errorCounter.incrementAndGet();
                                reconnectDelay = 1000;
                            }
                        } finally {
                            dispose();
                        }
                    }
                }
            };
            worker.start();
        }

        public void shutdown() {
            done.set(true);
            dispose();
        }

        public void waitForShutdown() throws InterruptedException {
            if (worker != null) {
                long start = System.currentTimeMillis();
                worker.join(6000);
                while (worker.isAlive()) {
                    if (allowWorkerInterrupt) {
                        System.out.println(name + " - worker did not shutdown quickly.. interrupting thread.");
                        worker.interrupt();
                        worker.join(1000);
                    } else {
                        System.out.println(name + " - worker did not shutdown quickly...");
                        worker.join(1000);
                    }
                }
                if (closeThread != null) {
                    long wait = 6000 - (System.currentTimeMillis() - start);
                    if (wait > 0) {
                        closeThread.join(wait);
                    }
                    while (closeThread.isAlive()) {
                        if (allowWorkerInterrupt) {
                            System.out.println(name + " closing thread did not shutdown quickly.. interrupting thread.");
                            closeThread.interrupt();
                            closeThread.join(1000);
                        } else {
                            System.out.println(name + " closing thread did not shutdown quickly...");
                            closeThread.join(1000);
                        }
                    }
                    closeThread.join();
                }
                worker = null;
            }
        }

        abstract String name();
    }

    class ConsumerClient extends JMSClient {
        private final int id;
        private final boolean drain;

        @Override
        public int id() {
            return id;
        }

        public String name() {
            return "consumer " + id;
        }

        public ConsumerClient(int id, boolean drain) {
            this.id = id;
            this.drain = drain;
        }

        public void execute() throws Exception {

            Session session = txSize == 0 ?
                    connection.createSession(false, jmsAckMode()) :
                    connection.createSession(true, Session.SESSION_TRANSACTED);

            // Draining a durable topic is easy.. just unsubscribe it!
            if (drain && durable) {
                loadStartRendezvous(this, session);
                session.unsubscribe(name);
                done.set(true);
                return;
            }

            MessageConsumer consumer = durable ?
                    session.createDurableSubscriber((Topic) destination(id), name, selector, noLocal)
                    :
                    session.createConsumer(destination(id), selector, noLocal);

            loadStartRendezvous(this, session);

            final int txCounter[] = new int[]{0};
            MessageListener listener = (msg) -> {
                try {
                    long latency = System.nanoTime() - msg.getLongProperty("ts");
                    updateMaxLatency(latency);
                    consumerCounter.incrementAndGet();
                    consumerSleep.apply(this);
                    if (session.getAcknowledgeMode() == Session.CLIENT_ACKNOWLEDGE) {
                        msg.acknowledge();
                    }

                    if (txSize != 0) {
                        txCounter[0] += 1;
                        if (txCounter[0] == txSize) {
                            session.commit();
                            txCounter[0] = 0;
                        }
                    }
                } catch (JMSException e) {
                    e.printStackTrace();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            };

            if (useMessageListener) {
                consumer.setMessageListener(listener);
                while (!done.get()) {
                    Thread.sleep(500);
                }
            } else {
                while (!done.get()) {
                    Message msg = consumer.receive(500);
                    if (msg != null) {
                        listener.onMessage(msg);
                    } else {
                        // commit early once we don't get anymore messages.
                        if (txSize != 0 && txCounter[0] > 0) {
                            session.commit();
                            txCounter[0] = 0;
                        }
                    }
                }
            }
            consumer.close();
        }

    }

    class ProducerClient extends JMSClient {

        private final int id;

        ProducerClient(int id) {
            this.id = id;
        }

        @Override
        public int id() {
            return id;
        }

        public String name() {
            return "producer " + id;
        }

        public void execute() throws Exception {

            Session session = txSize == 0 ?
                    connection.createSession(false, jmsAckMode())
                    :
                    connection.createSession(true, Session.SESSION_TRANSACTED);

            MessageProducer producer = session.createProducer(destination(id));
            producer.setDeliveryMode(persistent ? DeliveryMode.PERSISTENT : DeliveryMode.NON_PERSISTENT);

            TextMessage msg = session.createTextMessage(body(name()));
            headersFor(id).forEach(x -> {
                try {
                    msg.setStringProperty(x.getKey(), x.getValue());
                } catch (JMSException e) {
                    throw new RuntimeException(e);
                }
            });

            loadStartRendezvous(this, session);

            final int txCounter[] = new int[]{0};
            while (!done.get()) {
                msg.setLongProperty("ts", System.nanoTime());
                sendMessage(producer, msg);
                producerCounter.incrementAndGet();

                producerSleep.apply(this);
                if (txSize != 0) {
                    txCounter[0] += 1;
                    if (txCounter[0] == txSize) {
                        session.commit();
                        txCounter[0] = 0;
                    }
                }
            }

        }
    }

    void sendMessage(MessageProducer producer, TextMessage msg) throws Exception {
        producer.send(msg);
    }

    String body(String name)
    {
        StringBuffer buffer = new StringBuffer(messageSize);
        buffer.append("Message from " + name + "\n");
        for (int i = buffer.length(); i < messageSize; i++) {
            buffer.append('a' + (i % 26));
        }
        String rc = buffer.toString();
        if (rc.length() > messageSize) {
            return rc.substring(0, messageSize);
        } else {
            return rc;
        }
    }


}
